impress-org/givewp
# Provide additional account validation during PayPal Commerce onboarding

Resolves #4982 

## Description

There are quite a few things to be checked when onboarding an account, as the account has to be ready — on the PayPal side — in a number of ways before it is ready to accept donations.

### Account Readiness
Some of these account details have the ability to put the plugin into an "almost ready" sort of state, where some things will work and others won't. This is problematic as it's hard to communicate to the user and is _very_ likely to increase support load. I think it's better to stick to a simple pass/fail system and point the user to PayPal in most scenarios. That is, if the account is completely ready, then we're good to go; if the account is _not completely_ ready, we tell them what's missing and instruct them to reach out to PayPal. This isn't passing the buck; there's literally nothing we can do to help them.

This PR also stores "account readiness" as part of the MerchantDetail model. This is to indicate, after an account has been connected, whether or not the account is fully ready. This is currently used to display an admin notice if the gateway is active and in live mode, but the account is not ready. It should also probably be used on the front-end to prevent donors from attempting to use the gateway.

### Status Errors

I also changed status errors to persist rather than just show once and go away, as well as provide the ability to manually refresh their account status. Now any errors will display beneath the connection area.

![image](https://user-images.githubusercontent.com/2024145/89474891-b0f35200-d73b-11ea-9b06-630011489782.png)

### Additional Notices

Another notice has been added in the event that the gateway is active, in live mode, but the account has not been connected. This is especially important since the admin has to onboard twice: once for test mode, and again for live mode.

## Affects

The PayPal Commerce onboarding flow.

## Pre-review Checklist




## Testing Instructions

In the `/src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php`, after line 305, apply one of the following to mock an invalid status response:
```php
$onBoardedData['payments_receivable']           = false;
$onBoardedData['products'][1]['vetting_status'] = 'DENIED';
$onBoardedData['capabilities'][0]['status']     = 'DENIED';
```

Otherwise, review the Acceptance Criteria for what to check.
